core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -295,22 +295,19 @@ private void writeVersionHint(int versionToWrite) {
     FileSystem fs = getFileSystem(versionHintFile, conf);
 
     try {
-      if (fs.exists(versionHintFile)) {
-        Path tempVersionHintFile = metadataPath("version-hint.temp" );
-        writeFileToPath(fs, tempVersionHintFile);
-        fs.delete(versionHintFile, false);
-        fs.rename(tempVersionHintFile, versionHintFile);
-      } else {
-        writeFileToPath(fs, versionHintFile);
-      }
+      Path tempVersionHintFile = metadataPath(UUID.randomUUID().toString() + "-version-hint.temp");
+      writeVersionToPath(fs, tempVersionHintFile, versionToWrite);
+      fs.delete(versionHintFile, false);
+      fs.rename(tempVersionHintFile, versionHintFile);
     } catch (IOException e) {
       LOG.warn("Failed to update version hint", e);
     }
   }
 
-  private void writeFileToPath(FileSystem fs, Path path) throws IOException{
-    FSDataOutputStream out = fs.create(path, true /* overwrite */);
-    out.write(String.valueOf(path).getBytes(StandardCharsets.UTF_8));
+  private void writeVersionToPath(FileSystem fs, Path path, int versionToWrite) throws IOException {
+    try (FSDataOutputStream out = fs.create(path, true /* overwrite */)) {
```

Comment:
I wouldn't set overwrite to true to ensure this doesn't break anything. (In practice it won't, but it should be much clearer with overwrite = false.)

Reply:
Thanks for the review. Done